content/blog/2025-03-11-Alive-Peephole-Optimizations

```diff
@@ -0,0 +1,62 @@
+
+This new version of Alive, Alive2, has essentially replaced Alive’s usefulness in the LLVM development sphere, but it is still in use today and the contributions of Alive were essential in facilitating its creation.
+
+## Criticisms/Concerns:
+The authors do a lot to justify the usefulness of their tool; the anecdote about finding bugs and then eventually verifying correctness in a real developer’s actual LLVM patch is particularly effective. Reading the article, one can’t help but feel bad for that poor embarrassed developer for having bugs caught in the first two editions of their PR, but it is a convincing argument to use the tool; certainly from a software development perspective, you’d rather use Alive in-house and catch those bugs yourself, rather than send your patch out into the world and have someone else embarrass you by catching the bugs in your code!
+
+We also think it makes a lot of sense to design Alive with LLVM in mind, to find an already thriving developer community and create a tool specifically for them, rather than create something radically new and different that people are then resistant to actually use in real life. Tools with a concrete audience in mind are usually better loved and more impactful overall than even the most groundbreaking tool that people resist using. One can argue that since Alive can be rather difficult to understand for people without an LLVM background, there is some loss in impact with broader audiences for the language, but this is not really a shortcoming.
+
+However, there were a lot of concerns raised in the online discussion about both the scope/usefulness and correctness of the DSL. For example, with the scope of Alive only covering Peephole optimizations, which are small, simple cases that can be easily hand-checked, the sentiment kept popping up: how useful is the tool, really? There was a lot of interest in trying to expand a verification tool like Alive to more complex optimizations. However, through our discussion in class, we came to the conclusion that this kind of expansion was difficult to realistically achieve, and there was some dissent that this level of optimization was useless; the evidence that the paper presents about catching bugs in real-life LLVM programs using their tool suggests its usefulness, even if it only focuses on these small local peephole optimizations.
+
+There were also some concerns about correctness in Alive’s verification procedures, since they rely on the use of SMT solvers, which have known bugs in them. However, again in in-class discussion, we couldn’t come up with any better solution; to paraphrase Prof. Sampson in our in-class discussion, “It seems to me that SMT solvers are the worst solution, besides all the other solutions.”
```

Comment:
You may want to link to one of those papers about SMT solver bugs to substantiate your claim here. (It could be the same link as the one you use below.)